loadrec.1
.Dd 27 December, 2018
.Dt loadrec 1
.Os
.Sh NAME
.Nm loadrec
.Nd CPU load recorder
.Sh SYNOPSIS
.Nm
.Fl h
.Nm
.Op Fl v
.Op Fl d Ar ival
.Op Fl p Ar ival
.Op Fl o Ar file
.Sh DESCRIPTION
The
.Nm
command performs a recording of the current load. The purpose is to
reproduce this load to test different
.Xr powerd 8
and
.Xr powerd++ 8
configurations under identical load conditions using
.Xr loadplay 1 .
.Ss ARGUMENTS
The following argument types can be given:
.Bl -tag -width indent
.It Ar ival
A time interval can be given in seconds or milliseconds.
.D1 Li s , Li ms
An interval without a unit is treated as milliseconds.
.It Ar file
A file name.
.El
.Ss OPTIONS
The following options are supported:
.Bl -tag -width indent
.It Fl h , -help
Show usage and exit.
.It Fl v , -verbose
Be verbose and produce initial diagnostics on
.Pa stderr .
.It Fl d , -duration Ar ival
The duration of the recording session, defaults to 30 seconds.
.It Fl p , -poll Ar ival
The polling interval to take load samples at, defaults to 25 milliseconds.
.It Fl o , -output Ar file
The output file to write the load to.
.El
.Sh USAGE NOTES
To create reproducible results set a fixed CPU frequency below the
threshold at which the turbo mode is activated. E.g. an
.Nm Intel(R) Core(TM) i7-4500U CPU
supports the following frequency settings:
.Bd -literal -offset 4m
> sysctl dev.cpu.0.freq_levels
dev.cpu.0.freq_levels: 2401/15000 2400/15000 2300/14088 2200/13340 2000/11888 1900/11184 1800/10495 1700/9680 1500/8372 1400/7738 1300/7119 1200/6511 1100/5789 900/4643 800/4090 768/3550
.Ed
.Pp
Supposedly the first mode, which is off by 1 MHz, invokes the turbo
mode. However all modes down to 1800 MHz actually invoke the turbo
mode for this model. The only way to determine this is by benchmarking
the steppings to find out that there is a huge performance step between
1700 and 1800 MHz and that all the modes above 1700 MHz show the exact
same performance (given similar thermal conditions).
.Pp
So in order to produce a usable measurement for this CPU the clock
needs to be set to 1700 MHz or lower (higher is better to be able
to record a wider range of loads):
.Bd -literal -offset 4m
# service powerd++ stop
Stopping powerdxx.
Waiting for PIDS: 63574.
# powerd++ -M1700
.Ed
.Pp
Run
.Nm
for a brief time to test it:
.Bd -literal -offset 4m
> loadrec -d.25s
hw.machine=amd64
hw.model=Intel(R) Core(TM) i7-4500U CPU @ 1.80GHz
hw.ncpu=4
hw.acpi.acline=1
dev.cpu.0.freq=768
dev.cpu.0.freq_levels=2401/15000 2400/15000 2300/14088 2200/13340 2000/11888 1900/11184 1800/10495 1700/9680 1500/8372 1400/7738 1300/7119 1200/6511 1100/5789 900/4643 800/4090 768/3550
0 57097728 0 8232192 594432 100574976 58932480 0 7773696 32256 99760896 56408064 0 8003328 115200 101972736 56011008 0 7926528 33024 102528768
25 2304 0 768 0 0 1536 0 1536 0 0 3072 0 0 0 0 2304 0 768 0 0
25 1536 0 768 0 0 1536 0 768 0 0 2304 0 0 0 0 1536 0 768 0 0
25 1536 0 768 0 0 2304 0 0 0 0 768 0 768 768 0 2304 0 0 0 0
25 1536 0 768 0 0 2304 0 0 0 0 2304 0 0 0 0 2304 0 0 0 0
25 2304 0 0 0 0 1536 0 768 0 0 2304 0 0 0 0 2304 0 0 0 0
25 3072 0 0 0 0 2304 0 768 0 0 3072 0 0 0 0 3072 0 0 0 0
25 768 0 1536 0 0 2304 0 0 0 0 1536 0 768 0 0 2304 0 0 0 0
25 2304 0 0 0 0 1536 0 0 768 0 2304 0 0 0 0 2304 0 0 0 0
25 1536 0 768 0 0 2304 0 0 0 0 2304 0 0 0 0 1536 0 768 0 0
25 1536 0 768 0 0 1536 0 768 0 0 2304 0 0 0 0 1536 0 768 0 0
.Ed
.Pp
Printing the load creates significant load itself, so for the actual
measurement the output should be written to a file. Create your workload
and start your measurement:
.Bd -literal -offset 4m
> loadrec -o video-session.load
.Ed
.Pp
On the example setup
.Nm
produces a load of 0.001 (i.e. 0.1%), so its effect on the measurement is negligible.
.Sh SEE ALSO
.Xr cpufreq 4 , Xr loadplay 1 , Xr powerd 8 , Xr powerd++ 8 , Xr sysctl 8
.Sh AUTHORS
Implementation and manual by
.An Dominic Fandrey Aq kami@freebsd.org


src/loadrec.cpp
/** \file
 * Implements a load recorder, useful for simulating loads to test
 * CPU clock daemons and settings.
 */

#include "Options.hpp"

#include "types.hpp"
#include "constants.hpp"
#include "errors.hpp"
#include "utility.hpp"
#include "clas.hpp"
#include "version.hpp"

#include "sys/sysctl.hpp"

#include <iostream>  /* std::cout, std::cerr */
#include <fstream>   /* std::ofstream */
#include <chrono>    /* std::chrono::steady_clock::now() */
#include <thread>    /* std::this_thread::sleep_until() */
#include <memory>    /* std::unique_ptr */

#include <sys/resource.h>  /* CPUSTATES */

/**
 * File local scope.
 */
namespace {

using nih::Option;
using nih::make_Options;

using constants::ACLINE;
using constants::FREQ;
using constants::FREQ_LEVELS;
using constants::CP_TIMES;

using types::ms;
using types::coreid_t;
using types::cptime_t;
using types::mhz_t;

using errors::Exit;
using errors::Exception;
using errors::fail;

using utility::to_value;
using utility::sprintf_safe;
using namespace utility::literals;

using clas::ival;

using sys::ctl::make_Sysctl;
using sys::ctl::make_Once;

using version::LOADREC_FEATURES;
using version::flag_t;
using namespace version::literals;

/**
 * The set of supported features.
 *
 * This value is stored in load recordings to allow loadplay to correctly
 * interpret the data.
 */
constexpr flag_t const FEATURES{
	1_FREQ_TRACKING
};

/**
 * The global state.
 */
struct {
	bool verbose{false};  /**< Verbosity flag. */
	ms duration{30000};   /**< Recording duration in ms. */
	ms interval{25};      /**< Recording sample interval in ms. */

	/**
	 * The output file stream to use if an outfilename is provided
	 * on the CLI.
	 */
	std::ofstream outfile{};

	/**
	 * A pointer to the stream to use for output, either std::cout
	 * or outfile.
	 */
	std::ostream * out = &std::cout;

	/**
	 * The user provided output file name.
	 */
	char const * outfilename{nullptr};

	/**
	 * The number of CPU cores/threads.
	 */
	sys::ctl::SysctlOnce<coreid_t, 2> const ncpu{1U, {CTL_HW, HW_NCPU}};
} g;

/**
 * An enum for command line parsing.
 */
enum class OE {
	USAGE,           /**< Print help */
	IVAL_DURATION,   /**< Set the duration of the recording */
	IVAL_POLL,       /**< Set polling interval */
	FILE_OUTPUT,     /**< Set output file */
	FILE_PID,        /**< Set PID file */
	FLAG_VERBOSE,    /**< Verbose output on stderr */
	OPT_UNKNOWN,     /**< Obligatory */
	OPT_NOOPT,       /**< Obligatory */
	OPT_DASH,        /**< Obligatory */
	OPT_LDASH,       /**< Obligatory */
	OPT_DONE         /**< Obligatory */
};

/**
 * The short usage string.
 */
char const * const USAGE = "[-hv] [-d ival] [-p ival] [-o file]";

/**
 * Definitions of command line options.
 */
Option<OE> const OPTIONS[]{
	{OE::USAGE,         'h', "help",     "",     "Show usage and exit"},
	{OE::FLAG_VERBOSE,  'v', "verbose",  "",     "Be verbose"},
	{OE::IVAL_DURATION, 'd', "duration", "ival", "The duration of the recording"},
	{OE::IVAL_POLL,     'p', "poll",     "ival", "The polling interval"},
	{OE::FILE_OUTPUT,   'o', "output",   "file", "Output to file"},
	{OE::FILE_PID,      'P', "pid",      "file", "Ignored"},
};

/**
 * Outputs the given message on stderr if g.verbose is set.
 *
 * @param msg
 *	The message to output
 */
inline void verbose(std::string const & msg) {
	if (g.verbose) {
		std::cerr << "loadrec: " << msg << '\n';
	}
}

/**
 * Set up output to the given file.
 */
void init() {
	if (g.outfilename) {
		g.outfile.open(g.outfilename);
		if (!g.outfile.good()) {
			fail(Exit::EWOPEN, errno,
			     "could not open file for writing: "_s + g.outfilename);
		}
		g.out = &g.outfile;
	}
}

/**
 * Parse command line arguments.
 *
 * @param argc,argv
 *	The command line arguments
 */
void read_args(int const argc, char const * const argv[]) {
	auto getopt = make_Options(argc, argv, USAGE, OPTIONS);

	while (true) switch (getopt()) {
	case OE::USAGE:
		std::cerr << getopt.usage();
		throw Exception{Exit::OK, 0, ""};
	case OE::FLAG_VERBOSE:
		g.verbose = true;
		break;
	case OE::IVAL_DURATION:
		g.duration = ival(getopt[1]);
		break;
	case OE::IVAL_POLL:
		g.interval = ival(getopt[1]);
		break;
	case OE::FILE_OUTPUT:
		g.outfilename = getopt[1];
		break;
	case OE::FILE_PID:
		break;
	case OE::OPT_UNKNOWN:
	case OE::OPT_NOOPT:
	case OE::OPT_DASH:
	case OE::OPT_LDASH:
		fail(Exit::ECLARG, 0, "unexpected command line argument: "_s +
		                      getopt[0] + "\n\n" + getopt.usage());
	case OE::OPT_DONE:
		return;
	}
}

/**
 * Print the sysctls
 */
void print_sysctls() {
	sys::ctl::Sysctl<> hw_acpi_acline;
	try {
		hw_acpi_acline = {ACLINE};
	} catch (sys::sc_error<sys::ctl::error>) {
		verbose("cannot read "_s + ACLINE);
	}
	*g.out << "%s=%ld\n"_fmt(LOADREC_FEATURES, FEATURES)
	       << "hw.machine=" << make_Sysctl(CTL_HW, HW_MACHINE).get<char>().get() << '\n'
	       << "hw.model=" << make_Sysctl(CTL_HW, HW_MODEL).get<char>().get() << '\n'
	       << "hw.ncpu=" << g.ncpu << '\n'
	       << ACLINE << '=' << make_Once(1U, hw_acpi_acline) << '\n';

	for (coreid_t i = 0; i < g.ncpu; ++i) {
		char mibname[40];
		sprintf_safe(mibname, FREQ, i);
		try {
			sys::ctl::Sysctl<> ctl{mibname};
			*g.out << mibname << '='
			       << make_Once(0, ctl) << '\n';
		} catch (sys::sc_error<sys::ctl::error> e) {
			verbose("cannot access sysctl: "_s + mibname);
			if (i == 0) {
				fail(Exit::ENOFREQ, e,
				     "at least the first CPU core must report its clock frequency");
			}
		}
		sprintf_safe(mibname, FREQ_LEVELS, i);
		try {
			sys::ctl::Sysctl<> ctl{mibname};
			*g.out << mibname << '='
			       << ctl.get<char>().get() << '\n';
		} catch (sys::sc_error<sys::ctl::error>) {
			/* do nada */
		}
	}
}

/**
 * Report the load frames.
 *
 * This prints the time in ms since the last frame and the cp_times
 * growth as a space separated list.
 */
void run() try {
	/*
	 * Setup cptimes buffer for two samples.
	 */
	sys::ctl::Sysctl<> const cp_times_ctl = {CP_TIMES};

	auto const columns = cp_times_ctl.size() / sizeof(cptime_t);
	auto cp_times = std::unique_ptr<cptime_t[]>(
	    new cptime_t[2 * columns]{});

	/*
	 * Setup clock frequency sources for each core.
	 */
	coreid_t const cores = columns / CPUSTATES;
	auto corefreqs = std::unique_ptr<sys::ctl::Sync<mhz_t, sys::ctl::Sysctl<>>[]>(
	    new sys::ctl::Sync<mhz_t, sys::ctl::Sysctl<>>[cores]{});

	for (coreid_t i = 0; i < cores; ++i) {
		char mibname[40];
		sprintf_safe(mibname, FREQ, i);
		try {
			corefreqs[i] = sys::ctl::Sysctl<>{mibname};
		} catch (sys::sc_error<sys::ctl::error> e) {
			if (i == 0) {
				fail(Exit::ENOFREQ, e,
				     "at least the first CPU core must report its clock frequency");
			}
			/* Fall back to previous clock provider. */
			corefreqs[i] = corefreqs[i - 1];
		}
	}

	/*
	 * Record freq and cptimes.
	 */
	auto time = std::chrono::steady_clock::now();
	auto last = time;
	auto const stop = time + g.duration;
	size_t sample = 0;
	/* Takes a sample and prints it, avoids duplicating code
	 * behind the loop. */
	auto const takeAndPrintSample = [&]() {
		cp_times_ctl.get(&cp_times[sample * columns],
		                 sizeof(cptime_t) * columns);
		*g.out << std::chrono::duration_cast<ms>(time - last).count();
		for (coreid_t i = 0; i < cores; ++i) {
			*g.out << ' ' << static_cast<mhz_t>(corefreqs[i]);
		}
		for (int i = 0; i < columns; ++i) {
			*g.out << ' '
			       << cp_times[sample * columns + i] -
			          cp_times[((sample + 1) % 2) * columns + i];
		}
	};
	while (time < stop) {
		takeAndPrintSample();
		*g.out << '\n';
		sample = (sample + 1) % 2;
		last = time;
		std::this_thread::sleep_until(time += g.interval);
	}
	takeAndPrintSample();
	*g.out << std::endl;
} catch (sys::sc_error<sys::ctl::error> e) {
	fail(Exit::ESYSCTL, e, "failed to access sysctl: "_s + CP_TIMES);
}

} /* namespace */


/**
 * Main routine, setup and execute daemon, print errors.
 *
 * @param argc,argv
 *	The command line arguments
 * @return
 *	An exit code
 * @see Exit
 */
int main(int argc, char * argv[]) {
	try {
		read_args(argc, argv);
		init();
		print_sysctls();
		run();
	} catch (Exception & e) {
		if (e.msg != "") {
			std::cerr << "loadrec: " << e.msg << '\n';
		}
		return to_value(e.exitcode);
	} catch (sys::sc_error<sys::ctl::error> e) {
		std::cerr << "loadrec: untreated sysctl failure: " << e.c_str() << '\n';
		throw;
	} catch (...) {
		std::cerr << "loadrec: untreated failure\n";
		throw;
	}
}

